### PR TITLE
Added feature in llvm-profdata merge to filter functions from the profile

### PR DESCRIPTION
**Pull request by huangjd**
`--function=<regex>` Include functions matching regex in the output
`--no-function=<regex>` Exclude functions matching regex from the output

If both are specified, `--no-function` has a higher precedence if a function name matches both filters

### CODE REVIEW

**Comment by huangjd** [comment]
Purpose of this feature is to manually exclude certain functions from the profile for investigating compilation bug caused by the profile, or using this as a workaround

**Comment by wlei-llvm** [comment]
Thanks for the change. 
IIUC, it only supports filtering top-level functions, wondering if it's also useful for excluding the inlined(nested children) functions? thinking of some "investigating compilation bug" scenario, it might be necessary to ensure all the function copies are not loaded with a profile.

File: llvm/tools/llvm-profdata/llvm-profdata.cpp [hunk 1]
```diff
@@ -760,6 +767,45 @@ static void mergeWriterContexts(WriterContext *Dst, WriterContext *Src) {
...
+static std::string
+getFuncName(const SampleProfileMap::value_type &Val) {
+  return Val.second.getContext().toString();
+}
+
+template <typename T>
+static void filterFunctions(T &ProfileMap) {
+  bool hasFilter = !FuncNameFilter.empty();
+  bool hasNegativeFilter = !FuncNameNegativeFilter.empty();
+  if (hasFilter || hasNegativeFilter) {
```
Comment (david-xl) [inline llvm/tools/llvm-profdata/llvm-profdata.cpp:784]: Use early return to avoid the nesting.

**Review by david-xl — COMMENTED** [review]
please also update llvm-profdata.rst file in CommandGuide.

**Review by david-xl — APPROVED** [review]

**Comment by huangjd** [comment]
> Thanks for the change. IIUC, it only supports filtering top-level functions, wondering if it's also useful for excluding the inlined(nested children) functions? thinking of some "investigating compilation bug" scenario, it might be necessary to ensure all the function copies are not loaded with a profile.

When specifying a positive match filter, this will basically strip all inlinees of a matched function because its inlinees are likely to have different names so they will not be matched. 

We will need another way to specify the filter for inlined functions, as it is probably not intuitive to make the logic of positive and negative filter unexpectedly different.

**Pull request merged**